### PR TITLE
Feature/modify hca human experiments endpoint

### PR DESCRIPTION
**Pull request by upendrakumbham**
The goal of this feature is: Modify HCA Human experiments endpoint("/json/experiments/hca/human") to accept multiple organism parts. 

Example:` http://localhost:8080/sc/json/experiments/hca/human?organismPart=lymph node&organismPart=skin`

Changed Solr query to accept multiple organism parts.

In this PR we have switched from normal query to streaming expressions to improve performance. Here are the details for the proof of performance boost.
Pivotal story references - https://www.pivotaltracker.com/story/show/172792515
                                            https://www.pivotaltracker.com/story/show/172101611

### CODE REVIEW

File: src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java
Comment (haideriqbal) [inline src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java:34]: it would be preferable to use `ImmutableSet<String>` instead of `Set<String>`

File: src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java
Comment (haideriqbal) [inline src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java:42]: We need the normalize flag to exist for the cases when we just pass ontology id `e.g. UBERON_0000029`

When the normalize flag is not set then the resulting solr query is not correctly formed, which in this case would be `q=characteristic_name:("organism_part") AND characteristic_value("UBERON_0000029")...` which would return empty result set as there is no field with `characteristic_value: UBERON_0000029`.

So with the normalizing flag set to `false` the resulting query would be `q=characteristic_name:("organism_part") AND characteristic_value:*UBERON_0000029 ....` which is what we need.

File: src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java
Comment (haideriqbal) [inline src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java:45]: The default number of rows would not give us the correct result in case of this query because the result set is quite big. For example, when I ran test case `returnAllExperimentsForEmptyCharacteristicValue` with default rows I got the result set of just `14 experiment accessions` whereas when I ran with previous row numbers `10000000` I got a result set of `150 experiment accessions` which is the correct number of experiment accessions we expect to get. This is the result when I ran it on our wwwdev solr instance.

File: src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java
Comment (haideriqbal) [inline src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java:54]: as mentioned above we would need the concatenation of `*` in the case when just `ontology_id` is passed

File: src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java
Comment (haideriqbal) [inline src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java:65]: Why did we go from normal solr query to using solr stream expressions?

I believe in this case the performance with simple solr query is much faster than using stream expressions. For example when I ran test case `returnAllExperimentsForEmptyCharacteristicValue` with stream expression my laptop took `59 secs` to finish the test case whereas with normal solr query it just took `19 secs`.

File: src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentService.java
Comment (haideriqbal) [inline src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentService.java:24]: Same suggestion as above. Make more use of `ImmutableSet<String>` instead of `Set<String>`

File: src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentService.java
Comment (haideriqbal) [inline src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentService.java:34]: Remove blank space.

File: src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentService.java
Comment (haideriqbal) [inline src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentService.java:26]: `experimentAccessions` is a throw away variable here. You can directly do 
`return hcaHumanExperimentDao.fetchExperimentAccessions(characteristicName, characteristicValues)
.stream()
....`

File: src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentsController.java
Comment (haideriqbal) [inline src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentsController.java:30]: I guess there is no need for logging this information

File: src/test/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDaoIT.java
Comment (haideriqbal) [inline src/test/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDaoIT.java:51]: We need to pass just `ontologyId` as this is what we will be getting from `Anatomogram` instead of complete url

File: src/test/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDaoIT.java
Comment (haideriqbal) [inline src/test/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDaoIT.java:83]: These tests are not necessary for our use case but good to be cautious 👍

File: src/test/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDaoIT.java
Comment (haideriqbal) [inline src/test/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDaoIT.java:34]: Please also add a test case which has multiple `organism_parts` in the set to check if the functionality is working correctly or not.

File: src/test/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentServiceTest.java
Comment (haideriqbal) [inline src/test/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentServiceTest.java:37]: very minor but the name of the method starts with a lower case letter 🕵️‍♂️

File: src/test/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentServiceTest.java
Comment (haideriqbal) [inline src/test/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentServiceTest.java:70]: If you move `when(...) then(...)` statements in the `setUp` method then you would not need to repeat this statement in multiple methods also the test case themselves would be much cleaner and simple.

**Review by haideriqbal — CHANGES_REQUESTED** [review]
Please look at the comments. There are few changes which are required otherwise good work 👍

**Comment by upendrakumbham** [comment]
Thanks for the review comments, Haider. I will have a look.

File: src/test/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDaoIT.java
Comment (upendrakumbham) [reply src/test/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDaoIT.java:34]: Added test case.

File: src/test/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDaoIT.java
Comment (upendrakumbham) [reply src/test/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDaoIT.java:51]: Got it. Fixed.

File: src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentsController.java
Comment (upendrakumbham) [reply src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentsController.java:30]: Removed.

File: src/test/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentServiceTest.java
Comment (upendrakumbham) [reply src/test/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentServiceTest.java:37]: Sorry. Silly mistake.

File: src/test/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentServiceTest.java
Comment (upendrakumbham) [reply src/test/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentServiceTest.java:70]: It's true if we are using the same mocking in multiple test cases. 
But we are not repeating mock scenarios, they are different from each other.
If we move them back to setUp(), all mock scenarios would execute for every test case. 
Let me know if you are not happy with this, I can move them back to setUP() as you're the owner of this code :).

File: src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentService.java
Comment (upendrakumbham) [reply src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentService.java:26]: Removed throw away variable.

File: src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentService.java
Comment (upendrakumbham) [reply src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentService.java:34]: The tool is giving this extra black space before and end of class. 
I am not sure how the tool is working in you're development IDE.
Please, let me know to fix it if required.

File: src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java
Comment (upendrakumbham) [reply src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java:65]: Why did we go from normal solr query to using solr stream expressions? - the approach suggested by Alfonso. We will check with him what he would suggest before changing.

File: src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java
Comment (upendrakumbham) [reply src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java:54]: Fixed it.

File: src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java
Comment (upendrakumbham) [reply src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java:45]: Fixed it.

File: src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java
Comment (upendrakumbham) [reply src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java:42]: Yes, it make sense. I fixed it.

File: src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java
Comment (haideriqbal) [inline src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java:69]: 🏅

File: src/test/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentServiceTest.java
Comment (haideriqbal) [reply src/test/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentServiceTest.java:70]: Ohh I missed it earlier you are right they are different for each test so makes sense 👍 
Regarding ownership of the code, we all are owners of it as a team 😄

File: src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentsController.java
Comment (haideriqbal) [inline src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentsController.java:37]: remove blank space, please

File: src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java
Comment (haideriqbal) [reply src/main/java/uk/ac/ebi/atlas/hcalandingpage/HcaHumanExperimentDao.java:66]: Regarding this, I guess once you have profiled what is the bottleneck in solr java code then we can decide which approach to take.

**Review by haideriqbal — CHANGES_REQUESTED** [review]
Good work! Let's see what comes out of solr profiling and then we can decide and merge this PR.

**Comment by haideriqbal** [comment]
Performance appears to have improved significantly with new sharding in the solr. Just as a last curious step, can we run the code with streaming expressions and see how it has improved? The last time we ran we were getting around ~50 sec time.

**Comment by upendrakumbham** [comment]
Hi @haideriqbal, I am trying to run a few integration tests on this branch to proof code performance, But I am getting compilation issues on this branch because of this class `'ExperimentJsonSerializer'`. This has been modified by you in the core project after this PR. I took the latest code from the core and rebuild it. But still, I am getting the same error. Could you please try to run those tests if you can. It would be very helpful.
-Thank you

**Comment by alfonsomunozpomer** [comment]
@upendrakumbham Can you please check out a “compatible” commit in `atlas-web-core`? I think the commit right before 8aacc2c9f250891821d4bb468ffe85f02b35e6e2 in the master branch will work. Please, give that a go and let us know.

**Comment by upendrakumbham** [comment]
`git checkout 6ddb790730e0126967d2d5d84c33d014c4401678` core master branch. All errors gone.

Below tests, I ran:
```
uk.ac.ebi.atlas.hcalandingpage.HcaHumanExperimentDaoIT:

3.82 spassed notEmptyForWhenValidOnotologyIdIsPassed()
1 m 10 s passed notEmptyWhenValidOntologyAnnotationAncestorsOneOfTheLabelIsPassed()
3.96 s passed notEmptyForCorrectCharacteristicType()
1 m 28 spassedreturnAllExperimentsForEmptyCharacteristicValue()
3.74 spassednotEmptyWhenValidOntologyAnnotationLabelIsPassed()
122 ms passed emptyIfNoExperimentsCanBeFound()
4.76 s passed returnMatchingExperimentsForMultipleCharacteristicValues()
1 m 20 s passed notEmptyWhenValidOntologyAnnotationAncestorsOneOfTheUrlIsPassed()

```

**Comment by upendrakumbham** [comment]
git checkout  e9cece92985aecfee59967f189f43eddcf34d098. Core master branch.

I got `setFactes` related compilation errors in the `genesearchdao` and `cellmedatadatadao.  As they are very few I just commented out and ran tests.

`uk.ac.ebi.atlas.hcalandingpage.HcaHumanExperimentDaoIT:
-----
2.44 s passednotEmptyForWhenValidOnotologyIdIsPassed()
1 m 27 s passednotEmptyWhenValidOntologyAnnotationAncestorsOneOfTheLabelIsPassed()
5.22 s passednotEmptyForCorrectCharacteristicType()
1 m 37 s passedreturnAllExperimentsForEmptyCharacteristicValue()
4.09 s passednotEmptyWhenValidOntologyAnnotationLabelIsPassed()
126 ms passedemptyIfNoExperimentsCanBeFound()
5.41 s passedreturnMatchingExperimentsForMultipleCharacteristicValues()
1 m 29 s passednotEmptyWhenValidOntologyAnnotationAncestorsOneOfTheUrlIsPassed()

**Comment by upendrakumbham** [comment]
Above one right before 8aacc2c9f250891821d4bb468ffe85f02b35e6e2 in the master branch.

**Comment by upendrakumbham** [comment]
@Haider- you can update your results if you have any.

**Comment by haideriqbal** [comment]
I am getting this test `notEmptyForCorrectCharacteristicType` failed. Can you please check!

Regarding timings stat, I will leave it up to you to judge as the stats on my system are dodgy atm. If you think this is faster now let's go with streaming expressions then.

**Comment by upendrakumbham** [comment]
Yes, we are getting more than one experiment_acession for the characteristic_value 'skin'. That might be the reason, at least for me the reason for failed tests. Modified assertion to fix this.

Ran tests using Solr normal query:
--
Here are the results:

```
 uk.ac.ebi.atlas.hcalandingpage.HcaHumanExperimentDaoIT
--
5.02 s passed notEmptyForWhenValidOnotologyIdIsPassed()
 6.11 s passed notEmptyForCorrectCharacteristicType()
 2 m 29 s passed returnAllExperimentsForEmptyCharacteristicValue()
117 ms passed. emptyIfNoExperimentsCanBeFound()
```

**Review by haideriqbal — APPROVED** [review]
Cool! Can you please fix the assertions and commit it. Then you can resolve the conflicts and merge the PR.

**Comment by upendrakumbham** [comment]
` 2 m 29 s passed returnAllExperimentsForEmptyCharacteristicValue()` this is clearly saying that compare to normal Solr query streaming expressions reduced half of the time. So It proved to us that streaming expression boosting performance. So we are going with streaming expressions finally.

**Comment by upendrakumbham** [comment]
> Cool! Can you please fix the assertions and commit it. Then you can resolve the conflicts and merge the PR.

Yes, I will.

**Comment by upendrakumbham** [comment]
This PR got merged with the local master and resolved all the conflicts manually. I have tested the code in the local and pushed it to the remote.
Verified Bamboo, got a green build. Finally, this story is closed :).

**Comment by upendrakumbham** [comment]
this PR has been merged to master trough the command line. So I have closed this PR.

**Pull request closed without merging**
